src/dataflow-types/src/client/controller.rs

```diff
@@ -229,20 +256,53 @@ where
                         .await?;
                 let client = RemoteClient::new(&service.addresses("controller"));
                 let client: Box<dyn ComputeClient<T>> = Box::new(client);
-                self.compute_mut(instance)
+                self.compute_mut(instance_id)
                     .unwrap()
-                    .add_replica("default".into(), client)
+                    .add_replica(replica_name, client)
                     .await;
             }
         }
 
         Ok(())
     }
+
+    /// Removes a replica from an instance, including its service in the
+    /// orchestrator.
+    pub async fn drop_replica(
+        &mut self,
+        instance_id: ComputeInstanceId,
+        replica_id: ReplicaId,
+        config: ConcreteComputeInstanceReplicaConfig,
+    ) -> Result<(), anyhow::Error> {
+        let replica_name = generate_replica_service_name(instance_id, replica_id);
+        if let ConcreteComputeInstanceReplicaConfig::Managed {
+            size_config: _size_config,
+        } = config
+        {
+            let OrchestratorConfig { orchestrator, .. } = &self.orchestrator;
+            orchestrator
+                .namespace("compute")
+                .drop_service(&replica_name)
+                .await?;
+        }
+        let mut compute = self.compute_mut(instance_id).unwrap();
+        compute.remove_replica(&replica_name);
+        Ok(())
+    }
+
+    /// Removes an instance from the orchestrator.
+    ///
+    /// # Panics
+    /// - If the identified `instance` still has active replicas.
     pub async fn drop_instance(
         &mut self,
         instance: ComputeInstanceId,
     ) -> Result<(), anyhow::Error> {
         if let Some(mut compute) = self.compute.remove(&instance) {
+            assert!(
+                compute.client.get_replica_identifiers().next().is_none(),
+                "cannot drop instances with provisioned replicas; call `drop_replica` first"
+            );
```

Comment:
I'm a bit torn about this `assert`, since we don't have many other assertions in `Controller`. Another take would be that dropping the instance automatically drop all replicas -- @frankmcsherry what's your take on this?

Reply:
It should probably return an `Err` variant, if we want the erroring behavior? I don't know if it should be CASCADE or not (I would say yes myself, because that is how most non-SQL things work). I think if this converts to an `Err` then it is fine to land and we can consider changing the behavior later.

Reply:
Just have this returning an error now